framework/src/modules/fee/module.ts

```diff
@@ -76,43 +114,60 @@ export class FeeModule extends BaseModule {
 			this._tokenID,
 		);
 		if (transaction.fee > balance) {
-			return {
-				status: VerifyStatus.FAIL,
-				error: new Error('Insufficient balance.'),
-			};
+			throw new Error(`Insufficient balance.`);
 		}
 
 		return { status: VerifyStatus.OK };
 	}
 
 	public async beforeCommandExecute(context: TransactionExecuteContext): Promise<void> {
-		const {
-			header: { generatorAddress },
-			transaction: { senderAddress },
-		} = context;
-		const minFee = BigInt(this._minFeePerByte * context.transaction.getBytes().length);
+		const { transaction } = context;
 		const methodContext = context.getMethodContext();
+		await this._tokenMethod.lock(
+			methodContext,
+			transaction.senderAddress,
+			this.name,
+			this._tokenID,
+			transaction.fee,
+		);
+		const minFee = BigInt(this._minFeePerByte * context.transaction.getBytes().length);
 
-		const isNative = this._tokenMethod.isNativeToken(this._tokenID);
-		if (isNative) {
-			await this._tokenMethod.burn(methodContext, senderAddress, this._tokenID, minFee);
-			await this._tokenMethod.transfer(
-				methodContext,
-				senderAddress,
-				generatorAddress,
-				this._tokenID,
-				context.transaction.fee - minFee,
-			);
-
-			return;
-		}
+		context.contextStore.set(CONTEXT_STORE_KEY_AVAILABLE_FEE, transaction.fee - minFee);
+	}
 
+	public async afterCommandExecute(context: TransactionExecuteContext): Promise<void> {
+		const { header, transaction } = context;
+		await this._tokenMethod.unlock(
+			context.getMethodContext(),
+			transaction.senderAddress,
+			this.name,
+			this._tokenID,
+			transaction.fee,
+		);
+		const availableFee = getContextStoreBigInt(
+			context.contextStore,
+			CONTEXT_STORE_KEY_AVAILABLE_FEE,
+		);
+		await this._tokenMethod.burn(
+			context.getMethodContext(),
+			transaction.senderAddress,
+			this._tokenID,
+			transaction.fee - availableFee,
+		);
 		await this._tokenMethod.transfer(
-			methodContext,
-			senderAddress,
-			generatorAddress,
+			context.getMethodContext(),
+			transaction.senderAddress,
+			header.generatorAddress,
 			this._tokenID,
-			context.transaction.fee,
+			availableFee,
 		);
+
+		this.events.get(GeneratorFeeProcessedEvent).log(context, {
+			burntAmount: transaction.fee - availableFee,
+			generatorAddress: header.generatorAddress,
+			generatorAmount: availableFee,
+			senderAddress: transaction.senderAddress,
+		});
+		context.contextStore.delete(CONTEXT_STORE_KEY_AVAILABLE_FEE);
```

Comment:
maybe to assign zero instead of deleting the key?either way is fine but looks better to keep the key

Reply:
i think deleting is slightly better to keep the state clean